lib/model/repositories/FirestoreList.dart

```diff
@@ -0,0 +1,140 @@
+import 'dart:async';
+
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:farmsmart_flutter/model/bloc/Transformer.dart';
+
+class _Strings {
+  static const separator = "/";
+}
+
+typedef PathProvider = Future<String> Function();
+typedef ObjectIdentifier<T> = String Function(T object);
+
+class FireStoreList<T> {
+  final Firestore firestore;
+  final ObjectTransformer<T, Map<String, dynamic>> toFirestoreTransformer;
+  final ObjectTransformer<DocumentSnapshot, T> fromFirestoreTransformer;
+  final String orderField;
+  final bool orderDecending;
+  PathProvider path;
+  String _currentPath;
+  ObjectIdentifier<T> resourceIdentifier;
+  StreamController<List<T>> _listController;
+  StreamSubscription<QuerySnapshot> _subscription;
+
+  FireStoreList(this.firestore, this.toFirestoreTransformer,
+      this.fromFirestoreTransformer, this.path, this.resourceIdentifier,
+      {this.orderField, this.orderDecending = false});
+
+  Future<T> add(T object) {
+    return path().then((collectionPath) {
+      final firestoreObject = toFirestoreTransformer.transform(from: object);
+      return firestore
+          .collection(collectionPath)
+          .add(firestoreObject)
+          .then((result) {
+        return result.get().then((document) {
+          return fromFirestoreTransformer.transform(from: document);
+        });
+      });
+    });
+  }
+
+  Future<bool> remove(T object) {
+    return path().then((collectionPath) {
+      final documentPath = resourceIdentifier(object);
+      return firestore.document(documentPath).delete().then((response) {
+        return true;
+      }, onError: (error) {
+        return false;
+      });
+    });
+  }
+
+  Stream<List<T>> stream() {
+    if (_listController == null) {
+      _listController = StreamController<List<T>>.broadcast();
+    }
+    renewSubscriptionIfNeeded();
+    return _listController.stream;
+  }
+
+  Future<List<T>> get() {
+    final emptyList = List<T>();
+    return path().then((collectionPath) {
+      if (orderField != null) {
+        return _orderedCollection(
+          collectionPath,
+          orderField,
+          descending: orderDecending,
+        ).getDocuments().then((snapshot) {
+          return _update(snapshot);
+        }, onError: (error) {
+          return emptyList;
+        });
+      } else {
+        return firestore.collection(collectionPath).getDocuments().then(
+            (snapshot) {
+          return _update(snapshot);
+        }, onError: (error) {
+          return emptyList;
+        });
+      }
+    });
+  }
+
+  void renewSubscriptionIfNeeded() {
+    path().then((collectionPath) {
+      if (collectionPath != _currentPath) {
+        _subscription?.cancel();
+        if (collectionPath != null) {
+          if (orderField != null) {
```

Comment:
This if/else could be much cleaner if you split it into 2 different private functions indicating with right name what exactly are they doing.

Reply:
👍

Reply:
done